content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md

```diff
@@ -0,0 +1,375 @@
+		}
+	}
+	close(done)
+	close(errChan)
+
+	tt := time.Since(st)
+	size := len(keys)
+	pt := tt / time.Duration(size)
+	log.Printf("GRPC took %v for %d requests with %d client(s) (%v per each).\n", tt, size, numClients, pt)
+}
+
+```
+
+## Benchmark results
+
+Tests send 300,000 requests to key/value stores. One uses `jsonrpc`, and the other uses `gRPC`. Both `jsonrpc` and `gRPC` code use only one TCP connection. And another interesting `gRPC` case with a single TCP connection but with multiple clients:
+
+|RPC|# of requests|# of clients|total time|per-request time|
+|---|---|---|---|:-:|
+|`jsonrpc`|300,000|1|8m7.270s|1.624ms|
+|gRPC|300,000|1|36.715s|122.383µs|
+|gRPC|300,000|100|7.167s|23.892µs|
+
+<br>
+And if compared on memory usage:
+
+|RPC|`jsonrpc`|gRPC|delta|
+|---|---------|----|:---:|
+|NsPerOp|487271046903|36716116701|-92.46%|
+|AllocsPerOp|32747687|25221256|-22.98%|
+|AllocedBytesPerOp|3182814152|1795122672|-43.60%|
+
+|RPC|`jsonrpc`|gRPC with 100 clients|delta|
+|---|---------|----|:---:|
+|NsPerOp|487271046903|7168591678|-98.53%|
+|AllocsPerOp|32747687|25230286|-22.96%|
+|AllocedBytesPerOp|3182814152|1795831944|-43.58%|
+
+|RPC|gRPC|gRPC with 100 clients|delta|
+|---|---------|----|:---:|
+|NsPerOp|36716116701|7168591678|-80.48%|
+|AllocsPerOp|25221256|25230286|+0.04%|
+|AllocedBytesPerOp|1795122672|1795831944|+0.04%|
+
+<br>
+As you see, gRPC is much faster and lighter than `jsonrpc`. Not only performant, but also gRPC is easier to reason about concurrency. HTTP/1.x requires multiple TCP connections for concurrent requests, while HTTP/2 can have multiple requests in-flight at the same time over one single TCP connection and still process them asynchronously. Tests above show that gRPC with multiple clients solely speeds up gRPC by 80%, without having to open multiple TCP connections.
+
+## How does etcd use gRPC?
+
+etcd already uses protocol buffers for all structured data and supports gRPC as an experimental feature. etcd team is actively working on gRPC integration to replace all etcd RPCs. Here's how different parts of the etcd codebase leverage gRPC and Protocol Buffers:
+
+```
+etcdserver/         // Each etcd member runs as an etcdserver.
+
+	etcdserverpb/   // All message, schema, service are defined
+	                // here in Protocol Buffer format.
+
+	api/
+		v3rpc/      // Implements etcd v3 RPC system based on
+					// gRPC and Protocol Buffers.
+
+main.go             // Start 'etcdserver' with grpc framework.
+
+```
+
+## etcd 2.0 vs 2.3+
+
+- [etcd v2.0](https://github.com/coreos/etcd/tree/release-2.0) uses Protocol Buffer but does not support gRPC.
+- [etcd v2.3+](https://github.com/coreos/etcd/tree/b1138a42a28e3458ffc738e51f916c2e0c10966c) uses Protocol Buffer with gRPC support.
+
+Here's performance comparison:
+
+```
+
+IN PROGRESS...
+
+Will add tests results and code...
+
+```
+
+## Interested?
+
+etcd is still in active development, while widely used in production. If interested, please try it out:
```

Comment:
Sure, I will push changes in the next commit as you suggested. Thanks for feedback!